Marlin/src/module/planner.cpp

```diff
@@ -1605,37 +1606,48 @@ void Planner::check_axes_activity() {
   }
 
   void Planner::unapply_leveling(xyz_pos_t &raw) {
+    if (!leveling_active) return;
 
-    if (leveling_active) {
+    #if ABL_PLANAR
 
-      #if ABL_PLANAR
+      matrix_3x3 inverse = matrix_3x3::transpose(bed_level_matrix);
 
-        matrix_3x3 inverse = matrix_3x3::transpose(bed_level_matrix);
+      xy_pos_t d = raw - level_fulcrum;
+      inverse.apply_rotation_xyz(d.x, d.y, raw.z);
+      raw = d + level_fulcrum;
 
-        xy_pos_t d = raw - level_fulcrum;
-        inverse.apply_rotation_xyz(d.x, d.y, raw.z);
-        raw = d + level_fulcrum;
+    #elif HAS_MESH
 
-      #elif HAS_MESH
+      #if ENABLED(MESH_BED_LEVELING)
+        const float z_correction_fixed = mbl.get_z_correction_fixed();
+      #else
+        constexpr float z_correction_fixed = 0.0f;
+      #endif
 
-        #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
-          const float fade_scaling_factor = fade_scaling_factor_for_z(raw.z);
-        #elif DISABLED(MESH_BED_LEVELING)
-          constexpr float fade_scaling_factor = 1.0;
+      float z_correction_fadable = 
+        #if ENABLED(MESH_BED_LEVELING)
+          mbl.get_z_correction_fadable(raw);
+        #elif ENABLED(AUTO_BED_LEVELING_UBL)
+          ubl.get_z_correction(raw);
+        #elif ENABLED(AUTO_BED_LEVELING_BILINEAR)
+          bbl.get_z_correction(raw);
         #endif
 
-        raw.z -= (
-          #if ENABLED(MESH_BED_LEVELING)
-            mbl.get_z(raw OPTARG(ENABLE_LEVELING_FADE_HEIGHT, fade_scaling_factor))
-          #elif ENABLED(AUTO_BED_LEVELING_UBL)
-            fade_scaling_factor ? fade_scaling_factor * ubl.get_z_correction(raw) : 0.0
-          #elif ENABLED(AUTO_BED_LEVELING_BILINEAR)
-            fade_scaling_factor ? fade_scaling_factor * bbl.get_z_correction(raw) : 0.0
-          #endif
-        );
+      const float z_full_fade = raw.z - z_correction_fixed;
+      const float z_no_fade = z_full_fade - z_correction_fadable;
```

Comment:
Let me see if I understand these values… 
- `raw.z` is the Z returned from Stepper, which includes the faded bed leveling correction plus the non-faded MBL Z offset.
- `z_full_fade` removes from that the **_non-faded_ MBL Z offset**.
- `z_no_fade` further subtracts the **_non-faded_ bed leveling correction**.

This is where I start to get lost. The value of `z_no_fade` would not seem to be very meaningful in itself, but I see it is only used for specific cases below. Since we didn't subtract the faded offset, but the full one, we have taken off too much of the bed leveling correction for 0 < fade_factor < 1. So… I'll run it through the debugger and use `M114 D` and keep staring at it until it makes sense….

Reply:
Yes to your definitions. In addition...

`z_full_fade` is what cartesian Z will be if `fade_scaling_factor == 1`. Not that we know `fade_scaling_factor`, but if we did and it was 1, then this is the cartesian Z.

`z_no_fade` ditto for `fade_scaling_factor == 0`.